### PR TITLE
Glm

### PR DESCRIPTION
**Pull request by bmorel**

### CODE REVIEW

**Comment by slipher** [comment]
Can you explain what are the advantages of this library?

Apparently type safety isn't one of them, as this compiles:
```
glm::vec4 a;
glm::vec3 b = a;
```

I looked at a few commits at random and didn't find anything that became nicer than before.

**Comment by bmorel** [comment]
> Can you explain what are the advantages of this library?

Documented.
Maintained.
More readable.
Very likely more performance, but I don't care very much.

> Apparently type safety isn't one of them, as this compiles:
> 
> ```
> glm::vec4 a;
> glm::vec3 b = a;
> ```

Good spot. Compiler does not warn me here, too, maybe it's a bug that can be fixed upstream.

> I looked at a few commits at random and didn't find anything that became nicer than before.

I prefer:

```
glm::vec3 targetPos = self->botMind->goal.getPos();
```

Instead of:
```
vec3_t targetPos;
self->botMind->goal.getPos( targetPos );
```

Or:
```
void bla( glm::vec3& bar )
{
glm::vec3 foo;
foo = pos + length * dir;
...
bar = foo;
}
```
Instead of:

```
void bla( vec3_t bar ) {
vec3_t foo;
VectorMA( pos, length, dir, foo );
...
VectorCopy( foo, bar );
}
```

There are many examples like that, but *what I really enjoy* is: it's documented.
Most of the times I see the q_math API, I *have* to go read the damn code, and that included the `VectorMA` which took me more than a few days to finally be able to remember the exact order of operations.
In more obvious cases like VectorCopy, VectorSet or VectorClear, but I still find those annoying to read. Well, if that was only the readability problem, I would do with it, but again, *there is 0 doc* for q_math API.
Also, it uses damn macros. [A lot of those](https://github.com/DaemonEngine/Daemon/blob/075128eaaf1f9eb7a177892ac3b186b5f16ff794/src/engine/qcommon/q_shared.h#L405-L432) ( and they're as documented as the usual: not at all):

```
#define DotProduct( x,y )            ( ( x )[ 0 ] * ( y )[ 0 ] + ( x )[ 1 ] * ( y )[ 1 ] + ( x )[ 2 ] * ( y )[ 2 ] )
#define VectorSubtract( a,b,c )      ( ( c )[ 0 ] = ( a )[ 0 ] - ( b )[ 0 ],( c )[ 1 ] = ( a )[ 1 ] - ( b )[ 1 ],( c )[ 2 ] = ( a )[ 2 ] - ( b )[ 2 ] )
#define VectorAdd( a,b,c )           ( ( c )[ 0 ] = ( a )[ 0 ] + ( b )[ 0 ],( c )[ 1 ] = ( a )[ 1 ] + ( b )[ 1 ],( c )[ 2 ] = ( a )[ 2 ] + ( b )[ 2 ] )
#define VectorCopy( a,b )            ( ( b )[ 0 ] = ( a )[ 0 ],( b )[ 1 ] = ( a )[ 1 ],( b )[ 2 ] = ( a )[ 2 ] )
#define VectorScale( v, s, o )       ( ( o )[ 0 ] = ( v )[ 0 ] * ( s ),( o )[ 1 ] = ( v )[ 1 ] * ( s ),( o )[ 2 ] = ( v )[ 2 ] * ( s ) )
/** Stands for MultiplyAdd: adding a vector "b" scaled by "s" to "v" and writing it to "o" */
#define VectorMA( v, s, b, o )       ( ( o )[ 0 ] = ( v )[ 0 ] + ( b )[ 0 ] * ( s ),( o )[ 1 ] = ( v )[ 1 ] + ( b )[ 1 ] * ( s ),( o )[ 2 ] = ( v )[ 2 ] + ( b )[ 2 ] * ( s ) )
#define VectorLerpTrem( f, s, e, r ) (( r )[ 0 ] = ( s )[ 0 ] + ( f ) * (( e )[ 0 ] - ( s )[ 0 ] ), \
                                      ( r )[ 1 ] = ( s )[ 1 ] + ( f ) * (( e )[ 1 ] - ( s )[ 1 ] ), \
                                      ( r )[ 2 ] = ( s )[ 2 ] + ( f ) * (( e )[ 2 ] - ( s )[ 2 ] ))

#define VectorClear( a )             ( ( a )[ 0 ] = ( a )[ 1 ] = ( a )[ 2 ] = 0 )
#define VectorNegate( a,b )          ( ( b )[ 0 ] = -( a )[ 0 ],( b )[ 1 ] = -( a )[ 1 ],( b )[ 2 ] = -( a )[ 2 ] )
#define VectorSet( v, x, y, z )      ( ( v )[ 0 ] = ( x ), ( v )[ 1 ] = ( y ), ( v )[ 2 ] = ( z ) )

#define Vector2Set( v, x, y )        ( ( v )[ 0 ] = ( x ),( v )[ 1 ] = ( y ) )
#define Vector2Copy( a,b )           ( ( b )[ 0 ] = ( a )[ 0 ],( b )[ 1 ] = ( a )[ 1 ] )
#define Vector2Subtract( a,b,c )     ( ( c )[ 0 ] = ( a )[ 0 ] - ( b )[ 0 ],( c )[ 1 ] = ( a )[ 1 ] - ( b )[ 1 ] )

#define Vector4Set( v, x, y, z, n )  ( ( v )[ 0 ] = ( x ),( v )[ 1 ] = ( y ),( v )[ 2 ] = ( z ),( v )[ 3 ] = ( n ) )
#define Vector4Copy( a,b )           ( ( b )[ 0 ] = ( a )[ 0 ],( b )[ 1 ] = ( a )[ 1 ],( b )[ 2 ] = ( a )[ 2 ],( b )[ 3 ] = ( a )[ 3 ] )
/** Stands for MultiplyAdd: adding a vector "b" scaled by "s" to "v" and writing it to "o" */
#define Vector4MA( v, s, b, o )      ( ( o )[ 0 ] = ( v )[ 0 ] + ( b )[ 0 ] * ( s ),( o )[ 1 ] = ( v )[ 1 ] + ( b )[ 1 ] * ( s ),( o )[ 2 ] = ( v )[ 2 ] + ( b )[ 2 ] * ( s ),( o )[ 3 ] = ( v )[ 3 ] + ( b )[ 3 ] * ( s ) )
#define Vector4Average( v, b, s, o ) ( ( o )[ 0 ] = ( ( v )[ 0 ] * ( 1 - ( s ) ) ) + ( ( b )[ 0 ] * ( s ) ),( o )[ 1 ] = ( ( v )[ 1 ] * ( 1 - ( s ) ) ) + ( ( b )[ 1 ] * ( s ) ),( o )[ 2 ] = ( ( v )[ 2 ] * ( 1 - ( s ) ) ) + ( ( b )[ 2 ] * ( s ) ),( o )[ 3 ] = ( ( v )[ 3 ] * ( 1 - ( s ) ) ) + ( ( b )[ 3 ] * ( s ) ) )

#define DotProduct4(x, y)            (( x )[ 0 ] * ( y )[ 0 ] + ( x )[ 1 ] * ( y )[ 1 ] + ( x )[ 2 ] * ( y )[ 2 ] + ( x )[ 3 ] * ( y )[ 3 ] )

#define SnapVector( v )              do { ( v )[ 0 ] = ( floor( ( v )[ 0 ] + 0.5f ) ); ( v )[ 1 ] = ( floor( ( v )[ 1 ] + 0.5f ) ); ( v )[ 2 ] = ( floor( ( v )[ 2 ] + 0.5f ) ); } while ( 0 )
```

Speaking of, do you consider macros to be typesafe?

**Comment by slipher** [comment]
Noting a couple of things discussed on IRC:
- Define `GLM_FORCE_EXPLICIT_CTOR`, to avoid the unwanted type conversions discussed above
- Get rid of all instances of the Math::Vec3 (etc.) library

**Comment by bmorel** [comment]
> Get rid of all instances of the Math::Vec3 (etc.) library

This one can, and should, be done in a separate PR, since it would improve the current situation even if the migration to glm never ends (which is doubtful, but no-one knows the future).
This would make the glm transition easier, too, so to me it looks like the better path to go.

File: src/cgame/cg_tutorial.cpp [hunk 1]
```diff
@@ -225,7 +225,7 @@ static void CG_AlienBuilderText( char *text, playerState_t *ps )
 	if ( ps->stats[ STAT_CLASS ] == PCL_ALIEN_BUILDER0_UPG )
 	{
 		Q_strcat( text, MAX_TUTORIAL_TEXT,
-		          va( _( "Press %s to spit\n" ),
+		          va( _( "Press %s to spit. Spits can slow humans and remove fire.\n" ),
```
Comment (illwieckz) [inline src/cgame/cg_tutorial.cpp:228]: This one is probably a leftover from another branch. :grin:

**Comment by bmorel** [comment]
This will be reworked a lot when #1770 will be merged.

**Pull request closed without merging**